active/0000-lazy-load-overview.md

```diff
@@ -0,0 +1,187 @@
+we need a new hook in Ember's router or router.js, ongoing work for this is being done at the moment to make the
+`getHandler` function asynchronous. Once that is done, we can call the loading-service similar to what we do in
+components and resolve with the name of the asset to be loaded based on route metadata. The following code is
+just an example and assumes we have the route metadata and the loadingService available.
+A final implementation of how this hook will work is out of the scope of this RFC.
+
+```
+Router.getHandler = function (routeName) {
+  let assetName = routerMetadata.getAssetFor(routeName);
+  if (lazyLoaderService.loadedLibraries["assetName"]) {
+    return this._super(...arguments);
+  }
+  return lazyLoadingService.load(assetName).then(()=> {
+    this._super(...arguments);
+  });
+};
+```
+
+### Route metadata
+
+The code above assumes that we have the routerMetadata available, the definition of this needs further discussion 
+outside of the scope of this RFC. For the purpose of this overview we simply assume that we have a way to map between routeNames and assetNames.
+
+## Engines
+
+Engines can simply rely on all of the parts above and provide nice conventions.
+
+## Dependencies and concatenation
+
+Sometimes an asset has a dependency on another asset. For example,
+[ember-cli-bundle-loader](https://github.com/MiguelMadero/ember-cli-bundle-loader/blob/master/tests/dummy/config/bundles.js)
+allows you to specify dependencies bettween what they call `bundles`, the `lazyLoaderService` can make sure that the
+dependencies are loaded in parallel, but evaluated in the correct order and not loaded more than once.
+
+Another optimization is to concatenate dependencies into a single asset to avoid multiple requests. Today we can solve
+it during the build process by using the same `outputFile`, for engines and app code we might need to override some of
+the build defaults to oveirrde the output file.
+
+## Pre-requisites
+
+[Serialize RFC](https://github.com/emberjs/rfcs/pull/120)
+
+# Drawbacks
+
+TODO: fill this in
+
+* Without proper static analysis, it exposes the consumers to errors if there are dependencies in the order things are loaded
+and dependencies between assets aren't correctly specified. 
+* The way to build different assets is inconsistent. 
+
+
+# Alternatives
+
+This section needs further discussion. 
+
+TODO: fill this in
+
+# Unresolved questions
+
+* Ember Inspector breaks the apps the lazy-load routes because it eagerly resolves them, which results in a default route being generated, taking precedence over the lazy loaded one. We can fix Ember Inspector and even provide information about assets related to each route once we generate the route info at compile time. However, this will require users taking a new version of the extension. Not sure if we can force an update to avoid breakign the apps or if we can avoid that behavior on the app itself. 
+
```

Comment:
[ ] Add info about LOG_RESOLVER flag to the list of `unresolved questions`. It seems like all the route related objects are eagerly resolved if the flag is on.